lib/smf/smf.c

```diff
@@ -199,98 +242,156 @@ void smf_set_initial(struct smf_ctx *ctx, const struct smf_state *init_state)
 		init_state = init_state->initial;
 	}
 #endif
+
 	internal->exit = false;
 	internal->terminate = false;
 	ctx->current = init_state;
 	ctx->previous = NULL;
 	ctx->terminate_val = 0;
 
-	if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT)) {
-		internal->new_state = false;
+	internal->entry = true;
+
+#ifdef CONFIG_SMF_ANCESTOR_SUPPORT
+	ctx->executing = init_state;
+	const struct smf_state *topmost = get_last_of(init_state);
 
-		if (smf_execute_ancestor_entry_actions(ctx, init_state)) {
+	/* Execute topmost state entry action, since smf_execute_all_entry_actions() doesn't*/
+	if (topmost->entry) {
+		topmost->entry(ctx);
+		if (internal->terminate) {
+			/* No need to continue if terminate was set */
 			return;
 		}
 	}
 
-	/* Now execute the initial state's entry action */
+	if (smf_execute_all_entry_actions(ctx, init_state, topmost)) {
+		/* No need to continue if terminate was set */
+		return;
+	}
+#else
+	/* execute entry action if it exists */
 	if (init_state->entry) {
 		init_state->entry(ctx);
 	}
+#endif
+	internal->entry = false;
 }
 
-void smf_set_state(struct smf_ctx *const ctx, const struct smf_state *target)
+void smf_set_state(struct smf_ctx *const ctx, const struct smf_state *new_state)
 {
-	struct internal_ctx * const internal = (void *) &ctx->internal;
+	struct internal_ctx *const internal = (void *)&ctx->internal;
+
+	if (new_state == NULL) {
+		LOG_ERR("new_state cannot be NULL");
+		return;
+	}
 
 	/*
-	 * It does not make sense to call set_state in an exit phase of a state
+	 * It does not make sense to call smf_set_state_uml in an exit phase of a state
 	 * since we are already in a transition; we would always ignore the
 	 * intended state to transition into.
 	 */
 	if (internal->exit) {
-		LOG_WRN("Calling %s from exit action", __func__);
+		LOG_ERR("Calling %s from exit action", __func__);
 		return;
 	}
 
+#if CONFIG_SMF_ANCESTOR_SUPPORT
```

Comment:
Switch to `if (IS_ENABLED(...))` if possible.

Reply:
Nope, not possible as we're conditionally compiling in `parent` and `initial` in `smf_state`.